src/Microsoft.IdentityModel.Validators/AadIssuerValidator/AadIssuerValidator.cs

```diff
@@ -180,18 +191,24 @@ internal string AadAuthorityV2
 
             try
             {
-                BaseConfigurationManager effectiveConfigurationManager = GetEffectiveConfigurationManager(securityToken);
-                if (validationParameters.RefreshBeforeValidation)
-                    effectiveConfigurationManager.RequestRefresh();
+                var isV2Issuer = IsV2Issuer(securityToken);
 
-                BaseConfiguration configuration = await effectiveConfigurationManager.GetBaseConfigurationAsync(CancellationToken.None).ConfigureAwait(false);
-                string aadIssuer = configuration.Issuer;
+                var effectiveConfigurationManager = GetEffectiveConfigurationManager(isV2Issuer);
+                var baseConfiguration = await GetBaseConfiguration(effectiveConfigurationManager, validationParameters).ConfigureAwait(false);
+                string aadIssuer = baseConfiguration.Issuer;
 
                 if (!validationParameters.ValidateWithLKG)
                 {
                     if (IsValidIssuer(aadIssuer, tenantId, issuer))
                     {
-                        effectiveConfigurationManager.LastKnownGoodConfiguration = new OpenIdConnectConfiguration() { Issuer = aadIssuer };
+                        // We need to update LKG on the self-managed config manager.
+                        // The LKG of the one coming from provider, should be handled by the provider after successful validation
+                        // unless it is only fetched for the issuer and it's different than the authority.
+                        if (_configurationManagerProvider == null ||
```

Comment:
not understanding the conditions here, why not set LastKnownGood always?

Reply:
The scenario tests in SAL cover this.

This is what I was trying to raise during our conversation hinting at a little design weakness in LKG. 

In case the authority is the same as the token, the metadata is already fetched before the AadIssuerValidator is called but the LKG is set by an outside scope only after we verify the whole token on it [link](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/c3487c20fc5431e51c85020205598f40546bf794/src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs#L662C16-L669C18). In this case, if we were to set the LKG here, we could affect the LKG integrity if there an error would occur with metadata and the token after the issuer was validated. That is why I have opted to respect that and workaround it with not setting the LKG for the case when the metadata is fetched outside of AadIssuerValidator.

Now, in case of the token version being different than the authority, the issuer validator needs to request a new metadata that is not handled anywhere else (and currently today wont be used for anything else) so nobody would set its LKG anywhere, therefore this needs to be set here.

This seemed like a pragmatic approach and both scenarios are covered in SAL scenario test cases.